### PR TITLE
feat: add resources and granular permission groups for reconciliation

### PR DESCRIPTION
**Pull request by kashif-m**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Described in #6524 

This PR includes below changes
- Extending resources enum (for specifying the resource being consumed from the dashboard)
- Extending granular permission groups for recon
- Populating `acl` field in the recon token

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Helps maintain and configure permissions for recon module granularly.

## How did you test it?
- Verify the generated recon token has `acl`
- `GET /verify_token` response to have `acl`

<details>
    <summary>Check /verify_token response</summary>

cURL

    curl --location --request GET 'http://localhost:8080/recon/verify_token' \
        --header 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiZGJjMzJmMmYtYmI5Ni00MDI0LTliYTUtNjhkOTc3MzM0N2U5IiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNzMxOTk2NjQ4Iiwicm9sZV9pZCI6Im9yZ19hZG1pbiIsImV4cCI6MTczMjE4MzU4Miwib3JnX2lkIjoib3JnX0htMmY5dERZM2lGN0ZRckZUYUxKIiwicHJvZmlsZV9pZCI6InByb19GY0NaN1c4MllZVmNXMU1DSjZIOCIsInRlbmFudF9pZCI6InB1YmxpYyIsImFjbCI6IntcInJlY29uX2ZpbGVzXCI6XCJSV1wiLFwicmVjb25fY29uZmlnXCI6XCJSV1wiLFwicmVjb25fdG9rZW5cIjpcIlJXXCIsXCJyZWNvbl9yZXBvcnRzXCI6XCJSV1wiLFwicmVjb25fdXBsb2FkXCI6XCJSV1wiLFwicnVuX3JlY29uXCI6XCJSV1wiLFwicmVjb25fYW5kX3NldHRsZW1lbnRfYW5hbHl0aWNzXCI6XCJSXCJ9In0.wvS97F0dXwY7Y5uKSk8PEqHOhoAAPlR0coX_kIGyM2I'

Response

    {
        "merchant_id": "merchant_1731996648",
        "user_email": "kashif2@juspay.in",
        "acl": "{\"recon_reports\":\"RW\",\"recon_files\":\"RW\",\"run_recon\":\"RW\",\"recon_and_settlement_analytics\":\"R\",\"recon_token\":\"RW\",\"recon_upload\":\"RW\",\"recon_config\":\"RW\"}"
    }

</details>

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/services/authorization/permission_groups.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/permission_groups.rs:55]: Create new parent variants and use them instead.

File: crates/router/src/services/authorization/permission_groups.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/permission_groups.rs:63]: This can be removed once the new parent variants are added.

File: crates/router/src/services/authorization/permission_groups.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/permission_groups.rs:191]: We can remove this once new variants are added.

File: crates/router/src/routes/recon.rs
Comment (ThisIsMani) [inline crates/router/src/routes/recon.rs:57]: Can we have separate resource for `ReconToken` and use it here. We can put that resource in both the groups.

If we have this permission here, use should also have `AccountView` group to access this API.

File: crates/router/src/routes/user.rs
Comment (ThisIsMani) [inline crates/router/src/routes/user.rs:500]: Here as well.

File: crates/router/src/services/authentication.rs
Comment (ThisIsMani) [inline crates/router/src/services/authentication.rs:98]: Can't we use `AuthenticationDataWithUser` which you've created earlier?

**Review by ThisIsMani — CHANGES_REQUESTED** [review]

File: crates/router/src/services/authentication.rs
Comment (kashif-m) [reply crates/router/src/services/authentication.rs:98]: No, we need `RoleInfo` here

File: crates/api_models/src/user.rs
Comment (ThisIsMani) [inline crates/api_models/src/user.rs:177]: Can we move this API to recon as well?

File: crates/common_enums/src/enums.rs
Comment (ThisIsMani) [inline crates/common_enums/src/enums.rs:2818]: De we need this anymore?

File: crates/common_enums/src/enums.rs
Comment (ThisIsMani) [inline crates/common_enums/src/enums.rs:2801]: Let's not create a new group for this. The `ReconToken` resource should be part of the other 2 groups.

**Review by ThisIsMani — COMMENTED** [review]
Please remove the new group and Recon parent.

File: crates/common_enums/src/enums.rs
Comment (kashif-m) [reply crates/common_enums/src/enums.rs:2818]: Have kept it for ReconToken permissions

File: crates/common_enums/src/enums.rs
Comment (kashif-m) [reply crates/common_enums/src/enums.rs:2801]: We will not be able to specify multiple permissions to HyperSwitch API resources, so I've added this extra permission.

**Review by sahkal — DISMISSED** [review]
core part LGTM!

File: crates/router/src/services/authorization/permissions.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/permissions.rs:87]: What is the purpose of `Write` here? What will be operations on `Reports` that requires `Write`?

File: crates/router/src/services/authorization/roles.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/roles.rs:88]: Can we move this function as well to the recon files. It's not role's responsibility of the role to give the recon acl, but it's other way around.

**Review by ThisIsMani — COMMENTED** [review]
Let me know once you get approvals from PMs regarding recon groups in roles. I'll approve.

**Comment by ThisIsMani** [comment]
Other than that, LGTM.

File: crates/router/src/services/authorization/roles.rs
Comment (kashif-m) [reply crates/router/src/services/authorization/roles.rs:88]: Instead of making a dangling fn, we can let it be inside the `RoleInfo` impl, since it relies only on the RoleInfo struct

File: crates/router/src/services/authorization/roles/predefined_roles.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/roles/predefined_roles.rs:89]: There are no View variants here.

File: crates/router/src/services/authorization/roles/predefined_roles.rs
Comment (ThisIsMani) [inline crates/router/src/services/authorization/roles/predefined_roles.rs:32]: Here as well.

**Review by ThisIsMani — COMMENTED** [review]
@kashif-m, can you add view variants as well where we added manage groups in roles. This will maintain consistency with other roles.

**Review by ThisIsMani — DISMISSED** [review]

**Review by sahkal — DISMISSED** [review]

**Review by ThisIsMani — DISMISSED** [review]

**Review by ShivanshMathurJuspay — DISMISSED** [review]

File: crates/router/src/core/recon.rs
Comment (SanchithHegde) [inline crates/router/src/core/recon.rs:21]: Do we need this annotation?

File: crates/router/src/core/recon.rs
Comment (SanchithHegde) [inline crates/router/src/core/recon.rs:177]: By using the `let _` binding, we're ignoring the error occurring here, we're not even logging it. Is that intentional?

File: crates/router/src/core/recon.rs
Comment (kashif-m) [reply crates/router/src/core/recon.rs:21]: Yes, this is needed as `auth_data` in this fn is used behind `email` feature only

File: crates/router/src/core/recon.rs
Comment (kashif-m) [reply crates/router/src/core/recon.rs:177]: Ignoring this error is intentional (it's for notifying the end user about recon activation for their account, which is not an API failure)

Comment (kashif-m) [reply crates/router/src/core/recon.rs]: We can choose to log the errors though

**Review by SanchithHegde — DISMISSED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Pull request merged**
